nvflare/private/fed/utils/fed_utils.py

```diff
@@ -501,3 +501,15 @@ def get_scope_prop(scope_name: str, key: str) -> Any:
     check_str("key", key)
     data_bus = DataBus()
     return data_bus.get_data(_scope_prop_key(scope_name, key))
+
+
+def get_job_launcher(job_meta: dict, fl_ctx: FLContext) -> dict:
```

Comment:
This fl_ctx could be a long-running one and not job-specific. The property FLContextKey.JOB_LAUNCHER may already exist in it. Before firing the event, you need to remove this property, otherwise you may get the wrong result. But this solution is not thread-safe: in case this function is called multiple times at the same time, then it won't work properly.

The cleanest way is to create a new fl_ctx from the engine and then use it for the event processing.

Reply:
Changed to use a new fl_ctx for get_job_launcher to solve the thread safe issue.